server/src/main/java/org/opensearch/index/translog/InternalTranslogManager.java

```diff
@@ -0,0 +1,344 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.translog;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.apache.lucene.store.AlreadyClosedException;
+import org.opensearch.common.util.concurrent.ReleasableLock;
+import org.opensearch.index.engine.EngineConfig;
+import org.opensearch.index.seqno.LocalCheckpointTracker;
+import org.opensearch.index.shard.ShardId;
+import org.opensearch.index.translog.listener.TranslogEventListener;
+
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.function.LongConsumer;
+import java.util.function.LongSupplier;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+/**
+ * The {@link TranslogManager} implementation capable of orchestrating all read/write {@link Translog} operations while
+ * interfacing with the {@link org.opensearch.index.engine.InternalEngine}
+ *
+ * @opensearch.internal
+ */
+public class InternalTranslogManager implements TranslogManager {
+
+    private final ReleasableLock readLock;
+    private final Runnable ensureOpen;
+    private final ShardId shardId;
+    private final Translog translog;
+    private final BiConsumer<String, Exception> failEngine;
+    private final Function<AlreadyClosedException, Boolean> failOnTragicEvent;
+    private final AtomicBoolean pendingTranslogRecovery = new AtomicBoolean(false);
+    private final TranslogEventListener translogEventListener;
+    private static final Logger logger = LogManager.getLogger(InternalTranslogManager.class);
+
+    public InternalTranslogManager(
+        EngineConfig engineConfig,
+        ShardId shardId,
+        ReleasableLock readLock,
+        Supplier<LocalCheckpointTracker> localCheckpointTrackerSupplier,
+        String translogUUID,
+        TranslogEventListener translogEventListener,
+        Runnable ensureOpen,
```

Comment:
Ahh, should have renamed better, 
`ensureOpen` -> `ensureEngineOpen`, sorry that wasn't clear
```
BiConsumer<String, Exception> failEngine,
Function<AlreadyClosedException, Boolean> failOnTragicEvent
```
The above two are callbacks into Engine to let it shut gracefully in case the translog manager hits an unexpected failure

Reply:
`ensureEngineOpen` -> introduce `EngineLifecycle`, very easy to understand the purpose if `InternalTranslogManager` needs to follow the engine lifecycle (open / closed / ...)? 

 
BiConsumer<String, Exception> failEngine,  Function<AlreadyClosedException, Boolean> failOnTragicEvent  -> the listener with well defined hooks would definitely introduce more clarity, `onEngineFailure`, `onEngineClosed`

Reply:
> ensureEngineOpen -> introduce EngineLifecycle, 

Are you suggesting a `EngineLifecycleListener` to be implemented by `InternalTranslogManager `

Reply:
Not implemented but passed to `InternalTranslogManager`, afaik `InternalTranslogManager` needs to know if engine is open, right?

Reply:
Ah I see the confusion, I didn't think about [listener here](https://github.com/opensearch-project/OpenSearch/pull/3638#discussion_r902946037) but more like a lifecycle implementation (w/o state), `EngineLifecycle` implemented by `Engine`, in this case `ensureOpen` (+ close(), ...) could become the part of it.

Reply:
That would also mean passing critical Engine Lifecycle controls outside the class(in this case over to Translog) which I would avoid. So based on what you suggested this is what I think seems a better fit

```
public interface LifecycleAware {

    /**
     * Checks to ensure if the component is an open state
     */
    void ensureOpen();
```

Reply:
Just a suggestion, `EngineLifecycleAware`? Since it may not be generalized to anything besides the `Engine` itself?